tmc-plugin-intellij/src/main/java/fi/helsinki/cs/tmc/intellij/ui/testresults/TestResultsPanel.java

```diff
@@ -1,111 +1,196 @@
+        progressBar.setVisible(true);
+
+        this.revalidate();
+        this.repaint();
+    }
+
+    private void createConstraints() {
+        GridBagConstraints gbc = new GridBagConstraints();
+        gbc.gridx = 0;
+        gbc.gridy = GridBagConstraints.RELATIVE;
+        gbc.fill = GridBagConstraints.HORIZONTAL;
+        gbc.anchor = GridBagConstraints.NORTHWEST;
+        gbc.weightx = 1.0;
+        gbc.weighty = 0.0; // don't stretch vertically
+        gbc.insets.top = MARGIN;
+        this.resultsListConstraints = gbc;
+    }
+
+    private void createTestRows(List<TestResult> testResults) {
+        for (TestResult result : testResults) {
+            String details = getDetails(result);
+            TestResultRow row = createTestRow(result, details);
+            resultsList.add(row, resultsListConstraints);
+        }
+    }
+
+    private String getDetails(TestResult result) {
+        List<String> detailsRows = getDetailsOrMessageRows(result);
+        if (detailsRows == null) {
+            return null;
+        }
+        return String.join("\n", detailsRows);
+    }
+
+    private List<String> getDetailsOrMessageRows(TestResult res) {
+        return res.getDetailedMessage().size() > 0 ? res.getDetailedMessage() : res.getException();
+    }
+
+    private TestResultRow createTestRow(TestResult result, String details) {
+        if (result.isSuccessful()) {
+            return TestResultRow.createSuccessfulTestRow(
+                    result.getName(),
+                    result.getMessage(),
+                    details);
+        } else {
+            return TestResultRow.createFailedTestRow(
+                    result.getName(),
+                    result.getMessage(),
+                    details);
+        }
+    }
+
+    private void createValidationRows(ValidationResult validationResult) {
+        Map<File, List<ValidationError>> failedFiles = validationResult.getValidationErrors();
+
+        for (Map.Entry<File, List<ValidationError>> failedFile : failedFiles.entrySet()) {
+            String filename = failedFile.getKey().toString();
+            List<ValidationError> errors = failedFile.getValue();
+            createValidationRowForEachError(filename, errors);
         }
+    }
+
+    private void createValidationRowForEachError(String filename, List<ValidationError> errors) {
 
-        bar.setMinimum(0);
-        bar.setMaximum(100);
-        bar.setBorderPainted(false);
-        bar.setStringPainted(true);
-        bar.setValue((int) (100 * ((double) success / results.size())));
-        basePanel.repaint();
+        for (ValidationError error : errors) {
+            String message = String.format(
+                    "Line %d:    %s",
+                    error.getLine(),
+                    error.getMessage());
+
+            TestResultRow row =
+                    TestResultRow.createValidationRow(filename, message, error.getSourceName());
```

Comment:
Once `filename` is a file you can call `filename.getPath()` (which does the same thing as `toString` but is clearer)